File: lib/agents/execution/ab-variant-agent.ts
/**
 * Member 3 — Sub-agent 2: A/B Variant Agent
 *
 * Skill ref: .claude/.agents/skills/ab-test-orchestrator/SKILL.md
 *
 * Responsibilities:
 *  - Scrape competitor Meta Ad Library for real ad messaging
 *  - Search HN for sentiment signals
 *  - For each copy angle from ContentAgent, produce a CampaignVariant with:
 *    - A single falsifiable hypothesis tied to a specific research signal
 *    - A defined success metric
 *    - The single variable being tested
 *    - Grounded signal pointers
 */

import { searchMetaAds } from '../../tools/meta-ads';
import { searchHN } from '../../tools/hn-algolia';
import { searchWeb } from '../../tools/serpapi';
import { GoogleGenAI } from '@google/genai';
import type {
  AgentContext,
  AgentOutput,
  AgentSource,
  CampaignVariant,
  ConfidenceLevel,
} from '../types';
import { scoreToLevel } from '../types';
import { buildContentParts } from '../gemini-utils';
import { computeSignalQualityPenalty, extractToolResults } from '../../tools/fallback';

const ai = new GoogleGenAI({ apiKey: process.env.GEMINI_API_KEY! });

export interface ABVariantOutput extends AgentOutput {
  artifactType: 'execution-plan';
  variants: CampaignVariant[];
}

export async function runABVariantAgent(ctx: AgentContext): Promise<ABVariantOutput> {
  const { query, product, competitor, priorContext, researchOutputs = [], images } = ctx;

  // ── Parallel tool fetch ───────────────────────────────────────────────────
  const [metaAdsResult, hnResult, webResult] = await Promise.allSettled([
    competitor ? searchMetaAds(competitor) : searchMetaAds(product),
    searchHN(`${product} ${competitor ?? ''} outreach messaging`),
    searchWeb(`${product} vs ${competitor ?? 'competitors'} buyer decision 2025`),
  ]);

  const sources: AgentSource[] = [];
  const rawContent: string[] = [];

  // Research findings as grounding
  if (researchOutputs.length > 0) {
    const researchSummary = researchOutputs
      .map(o => `[${o.domain}] FACTS: ${o.facts.slice(0, 3).join('; ')} | INTERPRETATION: ${o.interpretation.slice(0, 2).join('; ')}`)
      .join('\n');
    rawContent.push(`[RESEARCH GROUNDING]\n${researchSummary}`);
  }

  if (metaAdsResult.status === 'fulfilled') {
    metaAdsResult.value.data.slice(0, 5).forEach(ad => {
      if (ad.ad_snapshot_url) {
        sources.push({ url: ad.ad_snapshot_url, title: `${ad.page_name} Ad`, timestamp: metaAdsResult.value.timestamp, tool: 'firecrawl' });
      }
      if (ad.ad_creative_body) rawContent.push(`[COMPETITOR AD] ${ad.page_name}: "${ad.ad_creative_body}"`);
    });
  }

  if (hnResult.status === 'fulfilled') {
    hnResult.value.data.slice(0, 4).forEach(p => {
      sources.push({ url: p.url, title: p.title, timestamp: p.created, tool: 'hn' });
      rawContent.push(`[HN SENTIMENT] ${p.title}`);
    });
  }

  if (webResult.status === 'fulfilled') {
    webResult.value.data.slice(0, 3).forEach(r => {
      sources.push({ url: r.url, title: r.title, timestamp: webResult.value.timestamp, tool: 'serpapi' });
      rawContent.push(`[BUYER DECISION] ${r.title}: ${r.snippet}`);
    });
  }

  // ── Gemini synthesis ──────────────────────────────────────────────────────
  const systemPrompt = `You are a specialist A/B test strategist for B2B SaaS outreach campaigns.

Your job is to produce 3 structured, hypothesis-driven message variants grounded in live signals.

Rules:
- Each variant tests ONE variable (angle/hook/frame) — not multiple things at once.
- Every hypothesis must be falsifiable: it predicts a specific outcome for a specific audience because of a specific signal.
- Variants must be meaningfully different: ROI, competitor gap, insight-led, no-headcount, pain-point-first are all distinct.
- Success metrics must be concrete and measurable (reply rate, meetings booked, etc.).
- Ground every hypothesis explicitly in a signal from the research or raw data.
- Output valid JSON matching the schema exactly.
${priorContext ? `\nPrior conversation context (build on this):\n${priorContext}` : ''}`;

  const userPrompt = `Query: "${query}"
Product: ${product}
${competitor ? `Competitor: ${competitor}` : ''}

Raw signals:
${rawContent.join('\n')}

Produce a JSON object with this exact shape:
{
  "variants": [
    {
      "id": string,                  // e.g. "V1-ROI"
      "angle": string,               // e.g. "ROI-focused"
      "hypothesis": string,          // falsifiable — e.g. "ROI messaging outperforms competitor-gap for Series B VP Sales because pricing-agent found 73% cite budget pressure"
      "successMetric": string,       // e.g. "reply rate > 4% within 72h"
      "variable": string,            // single thing being tested, e.g. "opening hook angle"
      "channels": {
        "email": {
          "subject": string,
          "body": string,            // 3-4 sentences, no placeholders
          "followUps": string[]      // 1-2 follow-up messages
        },
        "linkedin": {
          "hook": string,            // first line — must create curiosity
          "post": string             // 3-5 sentences
        }
      },
      "groundedSignals": string[]    // 2-3 direct quotes or paraphrases from raw signals above
    }
  ],
  "facts": string[],
  "interpretation": string[],
  "confidenceScore": number
}`;

  let parsed: any = {};
  try {
    const response = await ai.models.generateContent({
      model: 'gemini-2.0-flash',
      contents: [{ role: 'user', parts: buildContentParts(userPrompt, images) }],
      config: { systemInstruction: systemPrompt, responseMimeType: 'application/json' },
    });
    const text = response.candidates?.[0]?.content?.parts?.[0]?.text ?? '{}';
    const clean = text.replace(/```json\s*/i, '').replace(/```\s*/i, '').replace(/\s*```$/i, '').trim();
    parsed = JSON.parse(clean);
  } catch {
    parsed = {
      variants: [
        {
          id: 'V1-ROI',
          angle: 'ROI-focused',
          hypothesis: `ROI messaging outperforms competitor-gap for VP Sales at Series B because budget pressure is the primary buying signal for ${product}`,
          successMetric: 'reply rate > 4% within 72h',
          variable: 'opening hook angle',
          channels: {
            email: { subject: `What if ${product} paid for itself in 30 days?`, body: `Most teams spend 60% of selling time on manual research. ${product} flips that ratio — your reps get live-sourced intelligence in minutes, not days.\n\nWorth a 20-minute call to show you the math?`, followUps: [`Following up — did the ROI angle land?`] },
            linkedin: { hook: `Your top reps are losing deals to intel lag.`, post: `${product} surfaces live competitor, pricing, and buyer signals before the call even starts. Happy to share what that looks like for a team your size.` },
          },
          groundedSignals: ['Budget pressure identified in research signals', 'Competitor ad messaging focuses on speed'],
        },
        {
          id: 'V2-COMPETITOR-GAP',
          angle: 'Competitor gap',
          hypothesis: `Highlighting a specific competitor weakness increases curiosity opens for buyers already evaluating alternatives`,
          successMetric: 'click-through on demo link > 8%',
          variable: 'competitive framing vs ROI framing',
          channels: {
            email: { subject: `What ${competitor ?? 'your current tool'} doesn't show you`, body: `${competitor ?? 'Most intelligence tools'} gives you static reports. ${product} pulls live signals — funding moves, job postings, ad shifts — in real time.\n\nCurious what that difference looks like for your team?`, followUps: [`Still thinking it over? Happy to share a side-by-side.`] },
            linkedin: { hook: `Static research is a competitive liability in 2025.`, post: `${product} vs ${competitor ?? 'legacy tools'}: one surfaces live signals before the call, one summarises what happened last quarter. Reach out if you want to see the gap in action.` },
          },
          groundedSignals: ['Competitor ads emphasise static reports', 'HN sentiment shows frustration with lagging intel'],
        },
        {
          id: 'V3-INSIGHT-LED',
          angle: 'Insight-led',
          hypothesis: `Opening with a specific, non-obvious insight about the prospect's market increases reply rates among analytical buyers`,
          successMetric: 'reply rate > 5% for VP-level prospects',
          variable: 'personalised insight hook vs generic value prop',
          channels: {
            email: { subject: `One signal your team is probably missing`, body: `I ran ${product} against your category this morning — three competitors increased ad spend in the last 30 days while pulling back on pricing pages. That's usually a pivot signal.\n\nWant me to walk you through what it means for your Q3 positioning?`, followUps: [`Wanted to check — did the market signal angle resonate?`] },
            linkedin: { hook: `I found something interesting about your market this morning.`, post: `Using ${product} I spotted a pattern: three of your top competitors shifted ad messaging this week. Curious what that might mean for your positioning? Happy to share the breakdown.` },
          },
          groundedSignals: ['Market trend agent detected competitor ad spend shifts', 'Positioning agent identified untapped messaging angles'],
        },
      ],
      facts: rawContent.slice(0, 3).map(s => s.replace(/^\[[^\]]+\]\s*/, '')).filter(s => s.length > 15),
      interpretation: ['A/B variant synthesis temporarily unavailable — default variants generated.'],
      confidenceScore: 0.4,
    };
  }

  const rawScore: number = typeof parsed.confidenceScore === 'number' ? parsed.confidenceScore : 0.65;
  const toolResults = extractToolResults([metaAdsResult, hnResult, webResult]);
  const confScore = Number.parseFloat((rawScore * computeSignalQualityPenalty(toolResults, 3)).toFixed(2));
  const confidence: ConfidenceLevel = scoreToLevel(confScore);

  return {
    agentId: 'ab-variant-agent',
    domain: 'execution-engine',
    artifactType: 'execution-plan',
    confidence,
    confidenceScore: confScore,
    facts: parsed.facts ?? [],
    interpretation: parsed.interpretation ?? [],
    sources,
    generatedAt: new Date().toISOString(),
    variants: (parsed.variants ?? []) as CampaignVariant[],
  };
}

File: backend/app/agents/execution/ab_variant_agent.py
"""A/B Variant Agent — hypothesis-driven message variants (Stage 2 sub-agent)."""

from __future__ import annotations

import asyncio
import re
from datetime import datetime, timezone
from typing import Any

from app.gemini import generate_json
from app.models import score_to_level
from app.tools.fallback import compute_signal_quality_penalty, extract_tool_results
from app.tools.hn_algolia import search_hn
from app.tools.meta_ads import search_meta_ads
from app.tools.serpapi import search_web


async def run_ab_variant_agent(ctx: dict[str, Any]) -> dict[str, Any]:
    query = ctx["query"]
    product = ctx["product"]
    competitor = ctx.get("competitor")
    prior_context = ctx.get("prior_context")
    research_outputs = ctx.get("research_outputs") or []

    meta_ads_result, hn_result, web_result = await asyncio.gather(
        search_meta_ads(competitor) if competitor else search_meta_ads(product),
        search_hn(f"{product} {competitor or ''} outreach messaging"),
        search_web(f"{product} vs {competitor or 'competitors'} buyer decision 2025"),
        return_exceptions=True,
    )

    sources: list[dict[str, Any]] = []
    raw_content: list[str] = []

    if research_outputs:
        research_summary = "\n".join(
            f"[{o.get('domain', 'research')}] FACTS: {'; '.join((o.get('facts') or [])[:3])} | "
            f"INTERPRETATION: {'; '.join((o.get('interpretation') or [])[:2])}"
            for o in research_outputs
        )
        raw_content.append(f"[RESEARCH GROUNDING]\n{research_summary}")

    if not isinstance(meta_ads_result, BaseException) and meta_ads_result:
        for ad in (meta_ads_result.get("data") or [])[:5]:
            snapshot_url = ad.get("adSnapshotUrl") or ad.get("ad_snapshot_url")
            if snapshot_url:
                sources.append(
                    {
                        "url": snapshot_url,
                        "title": f"{ad.get('pageName') or ad.get('page_name', 'Competitor')} Ad",
                        "timestamp": meta_ads_result.get("timestamp", ""),
                        "tool": "firecrawl",
                    }
                )
            body = ad.get("adCreativeBody") or ad.get("ad_creative_body")
            if body:
                raw_content.append(
                    f"[COMPETITOR AD] {ad.get('pageName') or ad.get('page_name', '')}: \"{body}\""
                )

    if not isinstance(hn_result, BaseException) and hn_result:
        for post in (hn_result.get("data") or [])[:4]:
            sources.append(
                {
                    "url": post.get("url", ""),
                    "title": post.get("title", ""),
                    "timestamp": post.get("created", hn_result.get("timestamp", "")),
                    "tool": "hn",
                }
            )
            raw_content.append(f"[HN SENTIMENT] {post.get('title', '')}")

    if not isinstance(web_result, BaseException) and web_result:
        for result in (web_result.get("data") or [])[:3]:
            sources.append(
                {
                    "url": result.get("url", ""),
                    "title": result.get("title", ""),
                    "timestamp": web_result.get("timestamp", ""),
                    "tool": "serpapi",
                }
            )
            raw_content.append(f"[BUYER DECISION] {result.get('title', '')}: {result.get('snippet', '')}")

    system_prompt = """You are a specialist A/B test strategist for B2B SaaS outreach campaigns.

Your job is to produce 3 structured, hypothesis-driven message variants grounded in live signals.

Rules:
- Each variant tests ONE variable (angle/hook/frame) — not multiple things at once.
- Every hypothesis must be falsifiable: it predicts a specific outcome for a specific audience because of a specific signal.
- Variants must be meaningfully different: ROI, competitor gap, insight-led, no-headcount, pain-point-first are all distinct.
- Success metrics must be concrete and measurable (reply rate, meetings booked, etc.).
- Ground every hypothesis explicitly in a signal from the research or raw data.
- Output valid JSON matching the schema exactly."""

    if prior_context:
        system_prompt += f"\n\nPrior conversation context (build on this):\n{prior_context}"

    user_prompt = f"""Query: "{query}"
Product: {product}
{f"Competitor: {competitor}" if competitor else ""}

Raw signals:
{chr(10).join(raw_content)}

Produce a JSON object with this exact shape:
{{
  "variants": [
    {{
      "id": string,
      "angle": string,
      "hypothesis": string,
      "successMetric": string,
      "variable": string,
      "channels": {{
        "email": {{
          "subject": string,
          "body": string,
          "followUps": string[]
        }},
        "linkedin": {{
          "hook": string,
          "post": string
        }}
      }},
      "groundedSignals": string[]
    }}
  ],
  "facts": string[],
  "interpretation": string[],
  "confidenceScore": number
}}"""

    comp_label = competitor or "your current tool"
    try:
        parsed = await generate_json(system_prompt, user_prompt, max_new_tokens=1800, temperature=0.25)
    except Exception:
        parsed = {
            "variants": [
                {
                    "id": "V1-ROI",
                    "angle": "ROI-focused",
                    "hypothesis": (
                        f"ROI messaging outperforms competitor-gap for VP Sales at Series B because "
                        f"budget pressure is the primary buying signal for {product}"
                    ),
                    "successMetric": "reply rate > 4% within 72h",
                    "variable": "opening hook angle",
                    "channels": {
                        "email": {
                            "subject": f"What if {product} paid for itself in 30 days?",
                            "body": (
                                f"Most teams spend 60% of selling time on manual research. {product} flips "
                                f"that ratio — your reps get live-sourced intelligence in minutes, not days.\n\n"
                                f"Worth a 20-minute call to show you the math?"
                            ),
                            "followUps": ["Following up — did the ROI angle land?"],
                        },
                        "linkedin": {
                            "hook": "Your top reps are losing deals to intel lag.",
                            "post": (
                                f"{product} surfaces live competitor, pricing, and buyer signals before "
                                f"the call even starts. Happy to share what that looks like for a team your size."
                            ),
                        },
                    },
                    "groundedSignals": [
                        "Budget pressure identified in research signals",
                        "Competitor ad messaging focuses on speed",
                    ],
                },
                {
                    "id": "V2-COMPETITOR-GAP",
                    "angle": "Competitor gap",
                    "hypothesis": (
                        "Highlighting a specific competitor weakness increases curiosity opens for buyers "
                        "already evaluating alternatives"
                    ),
                    "successMetric": "click-through on demo link > 8%",
                    "variable": "competitive framing vs ROI framing",
                    "channels": {
                        "email": {
                            "subject": f"What {comp_label} doesn't show you",
                            "body": (
                                f"{comp_label} gives you static reports. {product} pulls live signals — "
                                f"funding moves, job postings, ad shifts — in real time.\n\n"
                                f"Curious what that difference looks like for your team?"
                            ),
                            "followUps": ["Still thinking it over? Happy to share a side-by-side."],
                        },
                        "linkedin": {
                            "hook": "Static research is a competitive liability in 2025.",
                            "post": (
                                f"{product} vs {comp_label}: one surfaces live signals before the call, "
                                f"one summarises what happened last quarter. Reach out if you want to see "
                                f"the gap in action."
                            ),
                        },
                    },
                    "groundedSignals": [
                        "Competitor ads emphasise static reports",
                        "HN sentiment shows frustration with lagging intel",
                    ],
                },
                {
                    "id": "V3-INSIGHT-LED",
                    "angle": "Insight-led",
                    "hypothesis": (
                        "Opening with a specific, non-obvious insight about the prospect's market "
                        "increases reply rates among analytical buyers"
                    ),
                    "successMetric": "reply rate > 5% for VP-level prospects",
                    "variable": "personalised insight hook vs generic value prop",
                    "channels": {
                        "email": {
                            "subject": "One signal your team is probably missing",
                            "body": (
                                f"I ran {product} against your category this morning — three competitors "
                                f"increased ad spend in the last 30 days while pulling back on pricing pages. "
                                f"That's usually a pivot signal.\n\n"
                                f"Want me to walk you through what it means for your Q3 positioning?"
                            ),
                            "followUps": ["Wanted to check — did the market signal angle resonate?"],
                        },
                        "linkedin": {
                            "hook": "I found something interesting about your market this morning.",
                            "post": (
                                f"Using {product} I spotted a pattern: three of your top competitors shifted "
                                f"ad messaging this week. Curious what that might mean for your positioning? "
                                f"Happy to share the breakdown."
                            ),
                        },
                    },
                    "groundedSignals": [
                        "Market trend agent detected competitor ad spend shifts",
                        "Positioning agent identified untapped messaging angles",
                    ],
                },
            ],
            "facts": [
                re.sub(r"^\[[^\]]+\]\s*", "", s)
                for s in raw_content[:3]
                if len(re.sub(r"^\[[^\]]+\]\s*", "", s)) > 15
            ],
            "interpretation": ["A/B variant synthesis temporarily unavailable — default variants generated."],
            "confidenceScore": 0.4,
        }

    raw_score = parsed.get("confidenceScore", 0.65)
    if not isinstance(raw_score, (int, float)):
        raw_score = 0.65
    tool_results = extract_tool_results([meta_ads_result, hn_result, web_result])
    conf_score = round(float(raw_score) * compute_signal_quality_penalty(tool_results, 3), 2)
    confidence = score_to_level(conf_score)

    return {
        "agentId": "ab-variant-agent",
        "domain": "execution-engine",
        "artifactType": "execution-plan",
        "confidence": confidence,
        "confidenceScore": conf_score,
        "facts": parsed.get("facts") or [],
        "interpretation": parsed.get("interpretation") or [],
        "sources": sources,
        "generatedAt": datetime.now(timezone.utc).isoformat(),
        "variants": parsed.get("variants") or [],
    }
